### PR TITLE
Feature expand s3 usage

### PR DESCRIPTION
**Pull request by teodorkostov-es**
- Add Deno project setup
- Use the bucket name only as part of the path
- Add an additional query parameters option

### CODE REVIEW

**Comment by jcs224** [comment]
FYI, not the maintainer of this package. But a note on improving your chance of getting the PR merged: Make separate PRs for each individual feature, instead of grouping them together. It's possible the author will like some of the changes, but not all of them.

**Comment by teodorkostov-es** [comment]
I am looking for a code review. Ready to discuss any remarks.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "deno.enable": true,
+  "deno.lint": true,
+  "deno.unstable": false,
+  "deno.config": "./tsconfig.json",
+}
```
Comment (dansalias) [inline .vscode/settings.json:6]: I prefer to keep the project editor-agnostic.

File: mod.ts [hunk 2]
```diff
@@ -3,8 +3,8 @@ import { Sha256, HmacSha256 } from 'https://deno.land/std/hash/sha256.ts'
 const NEWLINE = '\n'
 
 interface GetSignedUrlOptions {
-  bucketName: string
-  objectPath: string
+  path: string
```
Comment (dansalias) [inline mod.ts:6]: I understand the motivation, but independent `bucketName` and `objectPath` parameters is less ambiguous.

File: mod.ts [hunk 3]
```diff
@@ -3,8 +3,8 @@ import { Sha256, HmacSha256 } from 'https://deno.land/std/hash/sha256.ts'
...
+  query?: Record<string, string | number>
```
Comment (dansalias) [inline mod.ts:7]: :+1:

File: mod.ts [hunk 4]
```diff
@@ -28,17 +28,16 @@ function hmacSha256Hex(key: string | ArrayBuffer, data: string): string {
...
 function isoDate(date: Date): string {
-  return `${date.toISOString().substring(0,19).replace(/[^\dT]/g, '')}Z`
+  return `${date.toISOString().substring(0, 19).replace(/[^\dT]/g, '')}Z`
 }
 
 function parseOptions(provided: GetSignedUrlOptions): Required<GetSignedUrlOptions> {
-  if (!provided.objectPath.startsWith('/')) {
-    provided.objectPath = '/' + provided.objectPath
-  }
+  const fixedPath = `/${provided.path}`.replaceAll('//', '/')
```
Comment (dansalias) [inline mod.ts:39]: Nice. Though a little more difficult to understand intent at first glance.

File: mod.ts [hunk 5]
```diff
@@ -98,7 +102,8 @@ function getSignatureKey(options: Required<GetSignedUrlOptions>): string {
 
 function getUrl(options: Required<GetSignedUrlOptions>, queryParameters: URLSearchParams, signature: string): string {
   queryParameters.set('X-Amz-Signature', signature)
-  return `https://${options.bucketName}.${options.endpoint}${options.objectPath}?${new URLSearchParams(queryParameters).toString()}`
+
+  return `https://${options.endpoint}${options.path}?${new URLSearchParams(queryParameters).toString()}`
```
Comment (dansalias) [inline mod.ts:106]: Happy to switch it to this way around.

File: test.ts [hunk 6]
```diff
@@ -18,13 +15,13 @@ Deno.test('creates a presigned URL', () => {
...
     [
-      'https://examplebucket.s3.amazonaws.com/test.txt',
+      'https://s3.amazonaws.com/examplebucket/test.txt',
       '?X-Amz-Algorithm=AWS4-HMAC-SHA256',
       '&X-Amz-Credential=AKIAIOSFODNN7EXAMPLE%2F20130524%2Fus-east-1%2Fs3%2Faws4_request',
       '&X-Amz-Date=20130524T000000Z',
       '&X-Amz-Expires=86400',
       '&X-Amz-SignedHeaders=host',
-      '&X-Amz-Signature=aeeed9bbccd4d02ee5c0109b86d86835f995330da4c265957d157751f604d404',
+      '&X-Amz-Signature=733255ef022bec3f2a8701cd61d4b371f3f28c9f193a1f02279211d48d5193d7',
```
Comment (dansalias) [inline test.ts:24]: The signature should match that provided by https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html#query-string-auth-v4-signing-example to easily verify correctness.

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+  "compilerOptions": {
+    "lib": [
+      "esnext",
+      "deno.window",
+    ]
+  }
+}
```
Comment (dansalias) [inline tsconfig.json:8]: I believe this is covered by the implied Deno tsconfig. https://deno.land/manual@v1.28.3/advanced/typescript/configuration#what-an-implied-tsconfigjson-looks-like

**Comment by dansalias** [comment]
@teodorkostov-es thanks for contributing! I've added some comments, I'm happy to merge in the addition of a `queryParams` option.

**Comment by dansalias** [comment]
@teodorkostov-es to make it easier I've isolated those changes in a new PR (still accredited to you - https://github.com/dansalias/aws_s3_presign/pull/5/files). Let me know if that serves your needs.

**Comment by teodorkostov-es** [comment]
Hey @dansalias. Thank you for the review. Let's lead the discussion here. And it's OK if the result is another PR.

File: .vscode/settings.json (hunk 1 above)
Comment (teodorkostov-es) [reply .vscode/settings.json:6]: No problem. Will have to be removed.

File: mod.ts (hunk 2 above)
Comment (teodorkostov-es) [reply mod.ts:6]: Yes. Having used S3 more I agree that these should be separate.

However, I would suggest to use the AWS naming and call them `bucket` and `key`. This way the parameter names will be the same across the various S3 clients and libraries.

File: mod.ts (hunk 4 above)
Comment (teodorkostov-es) [reply mod.ts:39]: I would argue that control statements are harder to read. Without them the flow of the code is somewhat "uninterrupted" and easier to read.

File: mod.ts (hunk 5 above)
Comment (teodorkostov-es) [reply mod.ts:106]: Having used S3 more over the past months I have found an issue with this. Check out [this blog post](https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/).

The current accepted style is the one you have used.

However, there are S3 solutions that do not support the new style. Also, running such a solution on premise requires the certificates for the instance to accept all subdomains (e.g. `*.minio.local`). This might be an issue. So running a [Minio](https://min.io/) instance on-premise would be way easier with the older _S3 path request style_.

So, to summarize, I believe that the new request style has to be the default. However, there should be a way to use the older path request style.

How would you suggest to implement it? For example, [`s3fs`](https://github.com/s3fs-fuse/s3fs-fuse) uses a parameter `use_path_request_style`.

File: test.ts (hunk 6 above)
Comment (teodorkostov-es) [reply test.ts:24]: I believe that this is related to the request style that is used.

**Comment by teodorkostov-es** [comment]
Added the proposed changes in #6 with the project commit style.

File: mod.ts (hunk 5 above)
Comment (dansalias) [reply mod.ts:106]: Thanks for the clarification! I like the `usePathRequestStyle` option you've added.

File: mod.ts (hunk 4 above)
Comment (dansalias) [reply mod.ts:39]: I can get onboard with that :)

File: mod.ts (hunk 2 above)
Comment (dansalias) [reply mod.ts:6]: Understood, yes your suggestion has parity with the AWS SDK so it's the better option.

File: tsconfig.json (hunk 7 above)
Comment (teodorkostov-es) [reply tsconfig.json:8]: True. Thanks.

File: test.ts (hunk 6 above)
Comment (teodorkostov-es) [reply test.ts:24]: Yes, the official test cases should work.

**Comment by teodorkostov-es** [comment]
All discussions seem to be resolved. Closing this PR in favor of #6.
